src/main/kotlin/com/exactpro/th2/codec/api/impl/ThreadSafeCodec.kt

```diff
@@ -16,45 +16,40 @@
 
 package com.exactpro.th2.codec.api.impl
 
-import com.exactpro.sf.common.messages.structures.IDictionaryStructure
 import com.exactpro.th2.codec.api.IPipelineCodec
-import com.exactpro.th2.codec.api.IPipelineCodecSettings
+import com.exactpro.th2.codec.api.IPipelineCodecFactory
 import com.exactpro.th2.common.grpc.MessageGroup
 import java.util.concurrent.ConcurrentHashMap
 import javax.annotation.concurrent.ThreadSafe
 
 @ThreadSafe
-class ThreadSafeCodec(
-    private val dictionary: IDictionaryStructure,
-    private val settings: IPipelineCodecSettings?,
-    private val codecFactory: () -> IPipelineCodec
-) : IPipelineCodec {
-    private val instances = ConcurrentHashMap<Thread, IPipelineCodec>()
+class ThreadSafeCodec(private val codecFactory: IPipelineCodecFactory) : IPipelineCodec {
+    private val instances = ConcurrentHashMap<Long, IPipelineCodec>()
     override val protocol = getInstance().protocol
 
-    override fun init(dictionary: IDictionaryStructure, settings: IPipelineCodecSettings?) = error("Codec is already initialized")
-
-    override fun encode(messageGroup: MessageGroup) = getInstance().run {
-        synchronized(this) {
-            encode(messageGroup)
+    override fun encode(messageGroup: MessageGroup) = getInstance().let { codec ->
+        synchronized(codec) {
+            codec.encode(messageGroup)
         }
     }
 
-    override fun decode(messageGroup: MessageGroup) = getInstance().run {
-        synchronized(this) {
-            decode(messageGroup)
+    override fun decode(messageGroup: MessageGroup) = getInstance().let { codec ->
+        synchronized(codec) {
+            codec.decode(messageGroup)
         }
     }
 
-    private fun getInstance() = instances.computeIfAbsent(Thread.currentThread()) {
-        codecFactory().apply {
-            init(dictionary, settings)
-        }
+    private fun getInstance() = instances.computeIfAbsent(Thread.currentThread().id) {
+        synchronized(codecFactory, codecFactory::createCodec)
```

Comment:
What do we want to achieve via synchronized on the codecFactory?

Reply:
We need to synchronize here because codec factory isn't thread-safe